p4include/psa.p4

```diff
@@ -29,69 +29,277 @@ limitations under the License.
-typedef bit<16> PortIdInHeader_t;
-typedef bit<32> MulticastGroupInHeader_t;
-typedef bit<16> CloneSessionIdInHeader_t;
-typedef bit<8>  ClassOfServiceInHeader_t;
-typedef bit<16> PacketLengthInHeader_t;
-typedef bit<16> EgressInstanceInHeader_t;
-typedef bit<64> TimestampInHeader_t;
+
+/* See the comments near the definition of PortIdUint_t for why these
+ * typedef definitions exist. */
+typedef bit<32> PortIdInHeaderUint_t;
+typedef bit<32> MulticastGroupInHeaderUint_t;
+typedef bit<16> CloneSessionIdInHeaderUint_t;
+typedef bit<8>  ClassOfServiceInHeaderUint_t;
+typedef bit<16> PacketLengthInHeaderUint_t;
+typedef bit<16> EgressInstanceInHeaderUint_t;
+typedef bit<64> TimestampInHeaderUint_t;
+
+@p4runtime_translation("p4.org/psa/v1/PortIdInHeader_t", 32)
+type  PortIdInHeaderUint_t         PortIdInHeader_t;
+type  MulticastGroupInHeaderUint_t MulticastGroupInHeader_t;
+type  CloneSessionIdInHeaderUint_t CloneSessionIdInHeader_t;
+@p4runtime_translation("p4.org/psa/v1/ClassOfServiceInHeader_t", 8)
+type  ClassOfServiceInHeaderUint_t ClassOfServiceInHeader_t;
+type  PacketLengthInHeaderUint_t   PacketLengthInHeader_t;
+type  EgressInstanceInHeaderUint_t EgressInstanceInHeader_t;
+type  TimestampInHeaderUint_t      TimestampInHeader_t;
 // END:Type_defns2
 
+/* The _int_to_header functions were written to convert a value of
+ * type <name>_t (a value INTernal to the data path) to a value of
+ * type <name>InHeader_t inside a header that will be sent to the CPU
+ * port.
+ *
+ * The _header_to_int functions were written to convert values in the
+ * opposite direction, typically for assigning a value in a header
+ * received from the CPU port, to a value you wish to use in the rest
+ * of your code.
+ *
+ * The reason that three casts are needed is that each of the original
+ * and target types is declared via P4_16 'type', so without a cast
+ * they can only be assigned to values of that identical type.  The
+ * first cast changes it from the original 'type' to a 'bit<W1>' value
+ * of the same bit width W1.  The second cast changes its bit width,
+ * either prepending 0s if it becomes wider, or discarding the most
+ * significant bits if it becomes narrower.  The third cast changes it
+ * from a 'bit<W2>' value to the final 'type', with the same width
+ * W2. */
+
+PortId_t psa_PortId_header_to_int (in PortIdInHeader_t x) {
+    return (PortId_t) (PortIdUint_t) (PortIdInHeaderUint_t) x;
+}
+MulticastGroup_t psa_MulticastGroup_header_to_int (in MulticastGroupInHeader_t x) {
+    return (MulticastGroup_t) (MulticastGroupUint_t) (MulticastGroupInHeaderUint_t) x;
+}
+CloneSessionId_t psa_CloneSessionId_header_to_int (in CloneSessionIdInHeader_t x) {
+    return (CloneSessionId_t) (CloneSessionIdUint_t) (CloneSessionIdInHeaderUint_t) x;
+}
+ClassOfService_t psa_ClassOfService_header_to_int (in ClassOfServiceInHeader_t x) {
+    return (ClassOfService_t) (ClassOfServiceUint_t) (ClassOfServiceInHeaderUint_t) x;
```

Comment:
If I am not mistaken, the input to this function is of type `ClassOfServiceInHeader_t` which is 8 bits, and the output from this function is of type `ClassOfService_t` which is also 8 bits. Can you show a use case of these functions?

Reply:
What you say is true if a PSA implementation chooses to define ClassOfService_t to be 8 bits wide, but they can choose any smaller width for their PSA implementation, too, e.g. 3 bits.

Reply:
There are example uses of one of these conversion functions in this example program: https://github.com/p4lang/p4-spec/blob/master/p4-16/psa/examples/psa-example-parser-error-handling.p4#L380